test/test_common/environment.cc

```diff
@@ -0,0 +1,24 @@
+#include "test/test_common/environment.h"
+
+#include "common/common/assert.h"
+
+namespace {
+
+std::string* getCheckedEnvVar(const std::string& var) {
```

Comment:
The use of `new` in a function scoped `static` is one way to safely build a global/static non-POD object as per https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables. The `std::string` construction occurs on first use rather than non-deterministically. The object is never destructed, putting aside concerns over the non-deterministic destructor invocation ordering.

Thinking back to our previous discussion on this, I think I wasn't as concerned about non-POD globals due to the well behaved threading in Envoy. I still think the lack of destructor ordering is a problem (e.g. what if the `std::string` or `std::unique_ptr` is destroyed before some object with a ref to it?).

I recall the interplay with leak detection and this allocation style was a concern. Chatting @jamessynge, we came to the conclusion that the reason this doesn't trigger the leak detection is there is a reference from a live variable (the function scoped static) at program exit.